homeassistant/helpers/device_registry.py

```diff
@@ -136,8 +159,12 @@ def async_get_or_create(
         device = self.async_get_device(identifiers, connections)
 
         if device is None:
-            device = DeviceEntry(is_new=True)
-            self.devices[device.id] = device
+            device = self.async_get_deleted_device(identifiers, connections)
+            if device is None:
+                device = DeviceEntry(is_new=True)
+                self.devices[device.id] = device
+            else:
+                self.async_restore_device(device.id)
```

Comment:
```suggestion
            if deleted_device is None:
                device = DeviceEntry(is_new=True)
            else:
                device = deleted_device.to_device_entry()  # this method should set `is_new=True` as well. 

            self.devices[device.id] = device
```

Alternative is that we just do


```python
device = DeviceEntry(is_new=True)
if deleted_device is not None:
    device.identifiers = deleted_device.identifiers
    device.connections = deleted_device.connections
```

Reply:
Can't do that, `DeviceEntry` is frozen.